ucp/application_context.py

```diff
@@ -0,0 +1,376 @@
+# Copyright (c) 2019, NVIDIA CORPORATION. All rights reserved.
+# See file LICENSE for terms.
+
+import asyncio
+import logging
+import socket
+import struct
+import uuid
+import weakref
+from functools import partial
+from os import close as close_fd, environ
+from random import randint
+
+import psutil
+
+from . import public_api, send_recv
+from ._libs import ucx_api
+from .exceptions import UCXCanceled, UCXError, UCXWarning
+
+
+def asyncio_handle_exception(loop, context):
+    msg = context.get("exception", context["message"])
+    if isinstance(msg, UCXCanceled):
+        log = logging.debug
+    elif isinstance(msg, UCXWarning):
+        log = logging.warning
+    else:
+        log = logging.error
+    log("Ignored except: %s %s" % (type(msg), msg))
+
+
+async def exchange_peer_info(ucp_endpoint, msg_tag, ctrl_tag, guarantee_msg_order):
+    """Help function that exchange endpoint information"""
+
+    my_info = struct.pack("QQi", msg_tag, ctrl_tag, guarantee_msg_order)
```

Comment:
I think this needs to be `QQ?`. `guarantee_msg_order` is a bint which, I think, should be `?` for boolean on the Python side

Reply:
As I can never remember the codes for these things, here's the [table]( https://docs.python.org/3.8/library/struct.html#format-characters ).

Should we be worried about [endianness]( https://docs.python.org/3.8/library/struct.html#byte-order-size-and-alignment )?

Reply:
> I think this needs to be `QQ?`. `guarantee_msg_order` is a bint which, I think, should be `?` for boolean on the Python side

I think `"i"` is correct, according to [this](https://cython.readthedocs.io/en/latest/src/userguide/language_basics.html#types) `bint` is a `int` interpreted as a Python boolean.

Reply:
> Should we be worried about [endianness](https://docs.python.org/3.8/library/struct.html#byte-order-size-and-alignment)?

I think by omitting byte order symbol, we get automatic byte-order swap if the two endpoints have different endianness.
https://stackoverflow.com/questions/27506474/how-to-byte-swap-a-32-bit-integer-in-python

Reply:
Interesting. So `struct` is somehow tracking endianness?